BO Migration Shop parameters > Contact > Stores - Add Store entity

tom-combet opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      |  Add Store entity
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Related to #10624.
| How to test?      | N/A
| Possible impacts? | None.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27388)
<!-- Reviewable:end -->

Commit:
Add Store entity

Commit:
Add missing annotations to constraints

Commit:
Use Symfony Email constraint

Progi1984 commented on src/Core/ConstraintValidator/TypedRegexValidator.php:
```diff
@@ -141,6 +141,8 @@ private function getPattern($type)
             case TypedRegex::TYPE_URL:
                 return $this->characterCleaner->cleanNonUnicodeSupport('/^[~:#,$%&_=\(\)\.\? \+\-@\/a-zA-Z0-9\pL\pS-]+$/u');
+            case TypedRegex::TYPE_COORDINATE:
+                return '/^\-?[0-9]{1,8}\.[0-9]{1,8}$/s';
```
Could you add an unit test `tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php` ?

Progi1984 commented on src/PrestaShopBundle/Entity/Repository/StoreLangRepository.php:
```diff
@@ -0,0 +1,33 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity\Repository;
```
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity\Repository;
```

Progi1984 commented on src/PrestaShopBundle/Entity/Repository/StoreRepository.php:
```diff
@@ -0,0 +1,33 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity\Repository;
```
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity\Repository;
```

Progi1984 commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,322 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity;
```
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity;
```

Progi1984 commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,322 @@
+        return $this->shops;
+    }
+
+    public function addStoreLang(StoreLang $storeLang)
```
```suggestion
    public function addStoreLang(StoreLang $storeLang): self
```

Progi1984 commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,322 @@
+        return $this;
+    }
+
+    public function removeStoreLang(StoreLang $storeLang)
+    {
+        $this->storeLangs->removeElement($storeLang);
```
```suggestion
    public function removeStoreLang(StoreLang $storeLang): self
    {
        $this->storeLangs->removeElement($storeLang);
        
        return $this;
```

Progi1984 commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,322 @@
+        $this->storeLangs->removeElement($storeLang);
+    }
+
+    public function getStoreLangs()
```
```suggestion
    public function getStoreLangs(): ArrayCollection
```

Progi1984 commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,322 @@
+        return $this;
+    }
+
+    public function getShops(): Collection
```
```suggestion
    public function getShops(): ArrayCollection
```

Progi1984 commented on src/PrestaShopBundle/Entity/StoreLang.php:
```diff
@@ -0,0 +1,169 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity;
```
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity;
```

Progi1984 reviewed (commented):
@Kaudaj Some feedbacks

Commit:
Add missing strict_types directive

Commit:
Add missing return values

Commit:
Fix PHPDoc for collections

matks commented on src/PrestaShopBundle/Entity/Repository/StoreLangRepository.php:
```diff
@@ -0,0 +1,35 @@
+
+use Doctrine\ORM\EntityRepository;
+
+class StoreLangRepository extends EntityRepository
```
Hello as the PR does not contain the code that will make use of this class I cannot see if it's a good implementation or not.

I know you want to split the PR into small bits but if I don't have the context I cannot provide feedback about whether or not this code is valid.

matks commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+use PrestaShopBundle\Entity\Repository\StoreRepository;
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
```
Hello as the PR does not contain the code that will make use of this class I cannot see if it's a good implementation or not.

I know you want to split the PR into small bits but if I don't have the context I cannot provide feedback about whether or not this code is valid.

However so far we have not used Doctrine entities for back-office listing pages. If you look at other back-office listing pages migrated to Symfony we do not use Doctrine entities. So this looks like a wrong path.

tom-combet replied:
Thought Doctrine migration was topical, my bad. I'll finish it and put it aside until then.

Commit:
Remove useless default option label

Commit:
Add test for coordinate type

Commit:
Make constraint parameter type accurate

Progi1984 commented on tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php:
```diff
@@ -270,55 +270,21 @@ public function testItFailsForIsbnTypeWhenInvalidCharacterGiven(string $invalidC
         $this->assertViolationIsRaised(new TypedRegex('isbn'), $invalidChar);
     }
 
-    /**
```
I believe you have removed a lot of unit tests.

tom-combet replied:
Still don't know what happened. Should be good now! 😅

atomiix commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+    private $stateId;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("post_code")
```
An entity should always be valid: https://sarvendev.com/2018/01/entity-should-always-be-valid/
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+    private $postcode;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("city_name")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+    private $city;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("coordinate")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+    private $latitude;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("coordinate")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+    private $longitude;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("phone_number")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/Store.php:
```diff
@@ -0,0 +1,325 @@
+    private $phone;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("phone_number")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/StoreLang.php:
```diff
@@ -0,0 +1,171 @@
+    private $lang;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("generic_name")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/StoreLang.php:
```diff
@@ -0,0 +1,171 @@
+    private $name;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("address")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/StoreLang.php:
```diff
@@ -0,0 +1,171 @@
+    private $address1;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("address")
```
```suggestion
```

atomiix commented on src/PrestaShopBundle/Entity/StoreLang.php:
```diff
@@ -0,0 +1,171 @@
+    private $hours;
+
+    /**
+     * @PrestaShopAssert\CleanHtml
```
```suggestion
```

atomiix reviewed (changes requested):
Entities should always be valid, the data validation should be done before.
Also, I think that `StoreLangRepository` & `StoreRepository` classes can be removed as they do not contain new methods.

Kaudaj commented:
@atomiix Ah I maybe misunderstood [the doc](https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/validation/#validation), I thought validation was necessary anyway. I'll fix that.

Commit:
Remove validation from entities

An entity should always be valid: https://sarvendev.com/2018/01/entity-should-always-be-valid/

Commit:
Remove useless repositories

Progi1984 reviewed (changes requested):
@Kaudaj Could you remove entities ?

Kaudaj commented:
@Progi1984 Uh, why? This is the whole purpose of this PR... 😆
Can't I leave the PR on hold until the migration integrates the entities? Otherwise, I close it but I don't see the point of just removing the code.

Progi1984 commented:
@tom-combet The migration to entities will begin not in days or weeks,  more like in months or years. I think you can close this PR and keep your branch.
